component/src/action_handler/actions/ics20/withdrawal.rs

```diff
@@ -3,21 +3,31 @@ use std::sync::Arc;
 use anyhow::Result;
 use async_trait::async_trait;
 use penumbra_storage::{State, StateTransaction};
-use penumbra_transaction::action::Ics20Withdrawal;
+use penumbra_transaction::{action::Ics20Withdrawal, Transaction};
+use tracing::instrument;
 
 use crate::action_handler::ActionHandler;
 
 #[async_trait]
 impl ActionHandler for Ics20Withdrawal {
-    fn check_tx_stateless(&self) -> anyhow::Result<()> {
-        todo!()
+    #[instrument(name = "ics20_withdrawal", skip(self, _context))]
+    fn check_stateless(&self, _context: Arc<Transaction>) -> Result<()> {
+        self.validate()
     }
 
-    async fn check_tx_stateful(&self, _state: Arc<State>) -> Result<()> {
-        todo!()
+    #[instrument(name = "ics20_withdrawal", skip(self, state, _context))]
+    async fn check_stateful(&self, state: Arc<State>, _context: Arc<Transaction>) -> Result<()> {
+        <penumbra_storage::State as crate::ibc::transfer::Ics20TransferReadExt>::withdrawal_check(
+            state.clone(),
+            self,
+        )
```

Comment:
Hmm, this is a little weird, why doesn't `state.withdrawal_check(self)` work here?

Reply:
Ah, the implementation of `withdrawal_check` doesn't take a `self` parameter currently (due to copy-pasta) and rust-analyzer suggested this and I went with it. Should be easy to fix.